File: python/src/aegis/resources/anomaly.py
"""Anomaly resource — V3 anomaly detection endpoints.

``POST /v3/anomaly/detect`` accepts a flexible set of fields:
``metric``, ``algorithm`` ("zscore" / "moving_average" / "iqr" /
"isolation_forest"), ``data_points`` (array of {timestamp, value}),
``threshold``. The smoke-test friendly signature accepts ``value`` +
``history`` and converts to data_points.

``GET /v3/anomaly/events`` returns ``{events: [...], total: int}``.
"""

from __future__ import annotations

import time
from typing import Any, Dict, List, Optional

from aegis.resources._base import AsyncResource, SyncResource


def _detect_body(
    metric: Optional[str] = None,
    *,
    algorithm: str = "zscore",
    value: Optional[float] = None,
    history: Optional[List[float]] = None,
    data_points: Optional[List[Dict[str, Any]]] = None,
    threshold: Optional[float] = None,
    **kwargs: Any,
) -> Dict[str, Any]:
    body: Dict[str, Any] = {"algorithm": algorithm}
    if metric is not None:
        body["metric"] = metric
    if data_points is not None:
        body["data_points"] = data_points
    elif history is not None:
        now = int(time.time())
        points = [
            {"timestamp": now - (len(history) - i) * 60, "value": float(v)}
            for i, v in enumerate(history)
        ]
        if value is not None:
            points.append({"timestamp": now, "value": float(value)})
        body["data_points"] = points
    if threshold is not None:
        body["threshold"] = threshold
    body.update(kwargs)
    return body


class SyncAnomaly(SyncResource):

    def algorithms(self) -> List[Dict[str, Any]]:
        data = self._get("/v3/anomaly/algorithms")
        return data if isinstance(data, list) else data.get("algorithms", [])

    def detect(self, metric: Optional[str] = None, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/v3/anomaly/detect", json=_detect_body(metric, **kwargs))

    def events(self, *, limit: int = 20, **params: Any) -> Dict[str, Any]:
        params.setdefault("limit", limit)
        return self._get("/v3/anomaly/events", params=params)

    def stats(self) -> Dict[str, Any]:
        return self._get("/v3/anomaly/stats")


class AsyncAnomaly(AsyncResource):

    async def algorithms(self) -> List[Dict[str, Any]]:
        data = await self._get("/v3/anomaly/algorithms")
        return data if isinstance(data, list) else data.get("algorithms", [])

    async def detect(self, metric: Optional[str] = None, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/v3/anomaly/detect", json=_detect_body(metric, **kwargs))

    async def events(self, *, limit: int = 20, **params: Any) -> Dict[str, Any]:
        params.setdefault("limit", limit)
        return await self._get("/v3/anomaly/events", params=params)

    async def stats(self) -> Dict[str, Any]:
        return await self._get("/v3/anomaly/stats")

File: go/anomaly.go
package aegis

import (
	"context"
	"time"

	"github.com/ai-aegis-yatav/aegis-sdk/go/internal"
)

// AnomalyService — V3 anomaly detection.
type AnomalyService struct {
	t *internal.Transport
}

func (s *AnomalyService) Algorithms(ctx context.Context) (any, error) {
	var resp any
	if err := s.t.Do(ctx, "GET", "/v3/anomaly/algorithms", nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

// Detect runs anomaly detection. Provide either DataPoints, or History +
// optional Value (smoke-test convenience — converted to data_points).
type AnomalyDetectArgs struct {
	Metric     string           `json:"metric,omitempty"`
	Algorithm  string           `json:"algorithm"`
	DataPoints []map[string]any `json:"data_points,omitempty"`
	History    []float64        `json:"-"`
	Value      *float64         `json:"-"`
	Threshold  *float64         `json:"threshold,omitempty"`
}

func (s *AnomalyService) Detect(ctx context.Context, args AnomalyDetectArgs) (map[string]any, error) {
	if args.Algorithm == "" {
		args.Algorithm = "zscore"
	}
	body := map[string]any{"algorithm": args.Algorithm}
	if args.Metric != "" {
		body["metric"] = args.Metric
	}
	if args.DataPoints != nil {
		body["data_points"] = args.DataPoints
	} else if args.History != nil {
		now := time.Now().Unix()
		points := make([]map[string]any, 0, len(args.History)+1)
		for i, v := range args.History {
			points = append(points, map[string]any{
				"timestamp": now - int64(len(args.History)-i)*60,
				"value":     v,
			})
		}
		if args.Value != nil {
			points = append(points, map[string]any{"timestamp": now, "value": *args.Value})
		}
		body["data_points"] = points
	}
	if args.Threshold != nil {
		body["threshold"] = *args.Threshold
	}
	var resp map[string]any
	if err := s.t.Do(ctx, "POST", "/v3/anomaly/detect", body, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (s *AnomalyService) Events(ctx context.Context) (map[string]any, error) {
	var resp map[string]any
	if err := s.t.Do(ctx, "GET", "/v3/anomaly/events", nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (s *AnomalyService) Stats(ctx context.Context) (map[string]any, error) {
	var resp map[string]any
	if err := s.t.Do(ctx, "GET", "/v3/anomaly/stats", nil, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}
